pytorch_lightning/plugins/collective/torch_collective.py

```diff
@@ -0,0 +1,95 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any, Optional, Union
+
+import torch
+import torch.distributed
+
+from pytorch_lightning.plugins.collective import Collective
+from pytorch_lightning.utilities import _TORCH_GREATER_EQUAL_1_8
+from pytorch_lightning.utilities.apply_func import apply_to_collection
+from pytorch_lightning.utilities.distributed import (
+    all_gather_ddp_if_available,
+    distributed_available,
+    ReduceOp,
+    sync_ddp_if_available,
+)
+from pytorch_lightning.utilities.types import _METRIC_COLLECTION
+
+
+class TorchCollective(Collective):
+    """Collective interface for DDP, DDPSpawn, DP and DDP2."""
+
+    def __init__(self, local_reduce=False):
+        """.. note::
+
+        DDP and DDPSpawn sync accross multiple nodes/devices, local_reduce = False
+        DP run reduce in on node, local_reduce = True
+        DDP2 behaves like DP in one node, local_reduce = True
+
+        local_reduce set in Plugins.setup() functions
+        """
+        self.local_reduce = local_reduce
+
+    def barrier(self, *args, **kwargs) -> None:
+        if not distributed_available():
+            return
+        if _TORCH_GREATER_EQUAL_1_8 and torch.distributed.get_backend() == "nccl":
+            torch.distributed.barrier(device_ids=self.determine_ddp_device_ids())
+        else:
+            torch.distributed.barrier()
+
+    def broadcast(self, obj: object, src: int = 0) -> object:
+        if not distributed_available():
+            return obj
+        return self.dist.broadcast(obj)
```

Comment:
copy/paste issue from plugin?

Reply:
This one is on purpose. As DP and DDP both will use this class as collective plugin, we want to cover both use cases?

DDP broadcast
```   
def broadcast(self, obj: object, src: int = 0) -> object:
        return self.dist.broadcast(obj)
```

DDPSpawn broadcast
```   
def broadcast(self, obj: object, src: int = 0) -> object:
        if not distributed_available():
            return obj
        return self.dist.broadcast(obj)
```

DP broadcast:
   ```
def broadcast(self, obj: object, src: int = 0) -> object:
        return obj

```

Reply:
ah i mean there's no `self.dist` defined here

Reply:
Ohh, yeah!.. I think we can deprecate LightningDistributed? It's only used by ddp and ddpspawn, and we can directly include the broadcast function here?

Reply:
yes we should. the deprecation for it should be in a separate PR, but it ought to be succeeded by the collectives here. its one of the sources of collectives mentioned in the issue 😄